### ISSUE TITLE
specify bokeh port for dashboard

### ISSUE THREAD

**Issue by jbusecke** [pangeo-data/pangeo-discussion#14]
Let me start of with a big THANKS. This stuff is amazing!
I have sucessfully launched a notebook using dask-mpi by using a custom port specified in the jupyterconfig file. I am still not able to connect to the dashboard though, suspecting that the default port 8787 might not be available on my hpc cluster. Is there a way to specify the bokeh-port to a custom value?

**Comment by mrocklin** [comment]
Currently the `dask-mpi` command line tool is pretty bare-bones and doesn't include options like this.  They're very easy to add though.

I'm curious, are you on a machine that actually has access to the cluster machines?  I've always had to use ssh-tunneling in order to view the diagnostics dashboard.

**Comment by jbusecke** [comment]
Sorry I did not describe my workflow well. I actually tunneled to the cluster. The original tunnel to 8787 and 8888 did not work for either of them. Once I modified the jupyter notebook to run on 9992, that part worked but the other one still threw an error.
This is what I did to tunnel to tiger (tiger.princeton.edu)

`ssh -N -L 9992:tiger-r4c4n1:9992 -L 8787:tiger-r4c4n1:8787 tiger`

I can connect to the notebook server (9992) nicely but for the other port I get this error from the ssh command

```
channel 10: open failed: connect failed: Connection refused
channel 11: open failed: connect failed: Connection refused
channel 10: open failed: connect failed: Connection refused
channel 10: open failed: connect failed: Connection refused
```

Is there a way to reroute 8787 to a different port on the cluster itself? Sorry if this is a dumb question, I am still learning how to get around in the HPC environment.

Here are the two scripts I am using:

1) The batch job that sets up the scheduler:
```
#SBATCH --exclusive
#SBATCH -N 2
#SBATCH --ntasks-per-node=16
#SBATCH -J dask
#SBATCH --time=00:30:00
#SBATCH --mail-user=jbusecke@princeton.edu
#SBATCH --mail-type=ALL

# SLURM Template for Princeton Tiger
# Scheduler: SLURM

# This writes a scheduler.json file into your home directory
# You can then connect with the following Python code
# >>> from dask.distributed import Client
# >>> client = Client(scheduler_file='~/scheduler.json')

source activate standard
export XDG_RUNTIME_DIR=""
rm -f scheduler.json
mpirun --n 8 dask-mpi --interface ib0
```

2) and a python script that starts the jupyter notebook

```
from dask.distributed import Client
client = Client(scheduler_file='scheduler.json')

import socket
host = client.run_on_scheduler(socket.gethostname)

def start_jlab(dask_scheduler):
    import subprocess
    proc = subprocess.Popen(['jupyter', 'notebook', '--ip', host, '--no-browser','--port', '9992'])
    dask_scheduler.jlab_proc = proc

client.run_on_scheduler(start_jlab)
print("ssh -N -L 9992:%s:9992 -L 8787:%s:8787 tiger" % (host, host))
```

I would really love the functionality of the dashboard to assess the performance of this system since I just started working with it.

**Comment by jbusecke** [comment]
Sorry I did not describe my workflow well. I actually tunneled to the cluster. The original tunnel to 8787 and 8888 did not work for either of them. Once I modified the jupyter notebook to run on 9992, that part worked but the other one still threw an error.
This is what I did to tunnel to tiger (tiger.princeton.edu)

`ssh -N -L 9992:tiger-r4c4n1:9992 -L 8787:tiger-r4c4n1:8787 tiger`

I can connect to the notebook server (9992) nicely but for the other port I get this error from the ssh command

```
channel 10: open failed: connect failed: Connection refused
channel 11: open failed: connect failed: Connection refused
channel 10: open failed: connect failed: Connection refused
channel 10: open failed: connect failed: Connection refused
```

Is there a way to reroute 8787 to a different port on the cluster itself? Sorry if this is a dumb question, I am still learning how to get around in the HPC environment.

Here are the two scripts I am using:

1) The batch job that sets up the scheduler:
```
#SBATCH --exclusive
#SBATCH -N 2
#SBATCH --ntasks-per-node=16
#SBATCH -J dask
#SBATCH --time=00:30:00
#SBATCH --mail-user=jbusecke@princeton.edu
#SBATCH --mail-type=ALL

# SLURM Template for Princeton Tiger
# Scheduler: SLURM

# This writes a scheduler.json file into your home directory
# You can then connect with the following Python code
# >>> from dask.distributed import Client
# >>> client = Client(scheduler_file='~/scheduler.json')

source activate standard
export XDG_RUNTIME_DIR=""
rm -f scheduler.json
mpirun --n 8 dask-mpi --interface ib0
```

2) and a python script that starts the jupyter notebook

```
from dask.distributed import Client
client = Client(scheduler_file='scheduler.json')

import socket
host = client.run_on_scheduler(socket.gethostname)

def start_jlab(dask_scheduler):
    import subprocess
    proc = subprocess.Popen(['jupyter', 'notebook', '--ip', host, '--no-browser','--port', '9992'])
    dask_scheduler.jlab_proc = proc

client.run_on_scheduler(start_jlab)
print("ssh -N -L 9992:%s:9992 -L 8787:%s:8787 tiger" % (host, host))
```

I would really love the functionality of the dashboard to assess the performance of this system since I just started working with it.

**Comment by mrocklin** [comment]
See https://github.com/dask/distributed/pull/1412

**Comment by jbusecke** [comment]
Thank you so much @mrocklin. 
I tried it with a new port and it seems like I misdiagnosed the problem. I still cant connect to the port. I set up a dask.distributed scheduler on another machine (smaller, no job scheduling) and it turns out that the problem was the same when using the `--interface 'ib0'` option. It worked fine, when I didnt use that option. Could that be the culprit?
 
Is there anything that the initiated jupyter process does differently from dask-mpi if it comes to the port? I am curious, because in any setup I can readily connect to the port occupied by the jupyter lab/notebook.
